Feature profile model

ChadNaps opened the pull request:
### What this PR does (required):
- Adds profile model to the server, allowing it to store the following information
    - First Name
    - Last Name
    - Availability
    - Profile Picture
    - Gender
    - Birth Date
    - Phone Number
    - Address
    - Profile Description

### Any information needed to test this feature (required):
- Feature not integrated into app and there is no currently existing testing suite

### Any issues with the current functionality (optional):
- As this model is untested, it likely contains bugs or missing features, such as methods, that may be required when being integrated in the final application

ChadNaps committed:
Update .gitignore

Removes WebStorm configs.

ChadNaps committed:
Update README.md

Updates contributors.

ChadNaps committed:
Add Profile Model boilerplate

ChadNaps committed:
Add schema types and structure

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,132 @@
+const bcrypt = require("bcryptjs");
+
+const profileSchema = new mongoose.Schema({
+  user_username: {
```
Our `User` model already contains the `username` field, we wouldn't want to duplicate it :)

ChadNaps replied:
I was attempting to use the [Subset Pattern](https://docs.mongodb.com/manual/tutorial/model-embedded-one-to-one-relationships-between-documents/) that I found on the MongoDB website here. Is it a good idea or should I just remove `user_username`?

moffatethan replied:
This would actually just be storing the username on the profile. This means we would need to query the database specifically by the username. That's not something I would like to do, `userId` would be better to reference here. 

Subset pattern is great for breaking up large collections which ultimately can be a performance benefit. User collection is not very large so we don't need to really worry about sub dividing our collections out. 

👍🏻

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,132 @@
+    type: String,
+    required: true,
+  },
+  first_name: {
```
Please use camelCase for field names - it's the standard naming convention in JS 

Ditto throughout!

ChadNaps replied:
Initially I was planning to use camelCase, but I saw in the User model that the method was camelCase, but the schema fields were using underscores. For consistency, should I change the user model schema to camelCase as well or should I just do  the Profile model a little differently?

moffatethan replied:
Sure Chad, if you could update that as well. From this point on we'll preference `camelCase` 👍🏻

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,132 @@
+  },
+});
+
+module.exports = Profile = mongoose.model("profile", profileSchema);
```
```suggestion
module.exports = Profile = mongoose.model("Profile", profileSchema);
```
Nit: The naming convention for model names is PascalCase 😄

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,132 @@
+    end: Date,
+    days: {
+      mon: {
+        from: {
```
Alternatively, you could use a subdocument to set ranges here([docs](https://mongoosejs.com/docs/subdocs.html#subdocuments))

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,132 @@
+    type: Buffer,
+  },
+  gender: {
+    type: String,
```
Optional: You can use an enum validator for the types of gender

```suggestion
    type: String,
    enum: ['male','female','other'],
    lowercase: true, 
```

rajivtitus reviewed (commented):
Great work! 💯 
Just some minor comments to be addressed and this should be good 👍🏽

rajivtitus commented on server/models/Profile.js:
```diff
@@ -0,0 +1,132 @@
+  },
+  availability: {
+    begin: Date,
+    end: Date,
```
As a bonus, it would be nice to see some validations as well (e.g. end > start) for date ranges.

ChadNaps committed:
Refactor Profile.js

According to suggestion https://github.com/ChadNaps/fun-kickoff-project/pull/1#discussion_r780476217. Also removes bcrypt, which isn't used.

ChadNaps committed:
Refactor Profile.js

Uses Subdocuments, done according to feedback: https://github.com/ChadNaps/fun-kickoff-project/pull/1#discussion_r780481604

ChadNaps committed:
Update Profile.js

Adds enum to gender field. Done according to feedback: https://github.com/ChadNaps/fun-kickoff-project/pull/1#discussion_r780482489

ChadNaps committed:
fixup! Refactor Profile.js

Forgot to add schemas into arrays.

ChadNaps committed:
Update .gitignore and package-lock.json

ChadNaps committed:
Add validation to availability

ChadNaps committed:
Minor refactor

Changed because of Prettier plugin.

moffatethan reviewed (changes requested):
Let's fix up the schema keys to camelCase, commented on some other comments. You can change the user model in this PR too with that quick fix. Thanks Chad!

ChadNaps committed:
Refactor Profile and User models

Changes snake_case to camelCase.

ChadNaps committed:
Merge remote-tracking branch 'origin/feature-profile-model' into feature-profile-model
